<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>itpseq.DataSet &#8212; iTP-Seq 0.0.1a3 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=5ecbeea2" />
    <link rel="stylesheet" type="text/css" href="../_static/basic.css?v=686e5160" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css?v=27fed22d" />
    <link rel="stylesheet" type="text/css" href="../_static/sg_gallery.css?v=d2d258e8" />
    <script src="../_static/documentation_options.js?v=56909cd4"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="itpseq-dataset">
<h1>itpseq.DataSet<a class="headerlink" href="#itpseq-dataset" title="Link to this heading">¶</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="itpseq.DataSet">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">itpseq.</span></span><span class="sig-name descname"><span class="pre">DataSet</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'.'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">result_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keys</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_labels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">tuple</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'noa'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_pattern</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aafile_pattern</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#itpseq.DataSet" title="Link to this definition">¶</a></dt>
<dd><p>Loads an iTP-Seq dataset and provides methods for analyzing and visualizing the data.</p>
<p>A DataSet object is constructed to handle iTP-Seq Samples with their respective Replicates.
By default, it infers the files to uses in the provided directory by looking for “<a href="#id1"><span class="problematic" id="id2">*</span></a>.processed.json” files produced
during the initial step of pre-processing and filtering the fastq files.
It uses the pattern of the file names to group the Replicates into a Sample, and to define which condition
is the reference in the DataSet (the Sample with name “noa” by default).</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="itpseq.DataSet.data_path">
<span class="sig-name descname"><span class="pre">data_path</span></span><a class="headerlink" href="#itpseq.DataSet.data_path" title="Link to this definition">¶</a></dt>
<dd><p>Path to the data directory containing the output files from the fastq pre-processing.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str or Path</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="itpseq.DataSet.result_path">
<span class="sig-name descname"><span class="pre">result_path</span></span><a class="headerlink" href="#itpseq.DataSet.result_path" title="Link to this definition">¶</a></dt>
<dd><p>Path to the directory where the results of the analysis will be saved.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str or Path</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="itpseq.DataSet.samples">
<span class="sig-name descname"><span class="pre">samples</span></span><a class="headerlink" href="#itpseq.DataSet.samples" title="Link to this definition">¶</a></dt>
<dd><p>Dictionary of Samples in the DataSet. By default, it is None and will be populated automatically.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="itpseq.DataSet.keys">
<span class="sig-name descname"><span class="pre">keys</span></span><a class="headerlink" href="#itpseq.DataSet.keys" title="Link to this definition">¶</a></dt>
<dd><p>Properties in the file name to use for identifying the reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="itpseq.DataSet.ref_labels">
<span class="sig-name descname"><span class="pre">ref_labels</span></span><a class="headerlink" href="#itpseq.DataSet.ref_labels" title="Link to this definition">¶</a></dt>
<dd><p>Specifies the reference: e.g. ‘noa’ or ((‘sample’, ‘noa’),)</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str or tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="itpseq.DataSet.cache_path">
<span class="sig-name descname"><span class="pre">cache_path</span></span><a class="headerlink" href="#itpseq.DataSet.cache_path" title="Link to this definition">¶</a></dt>
<dd><p>Path used to cache intermediate results. By default, this creates a subdirectory called “cache” in the result_path directory.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str or Path</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="itpseq.DataSet.file_pattern">
<span class="sig-name descname"><span class="pre">file_pattern</span></span><a class="headerlink" href="#itpseq.DataSet.file_pattern" title="Link to this definition">¶</a></dt>
<dd><p>Regex pattern used to identify the sample files in the data_path directory.
If None, defaults to <cite>r’(?P&lt;lib_type&gt;[^_]+)_(?P&lt;sample&gt;[^_d]+)(?P&lt;replicate&gt;d+).processed.json’</cite>
which matches files like nnn15_noa1.processed.json, nnn15_tcx2.processed.json, etc.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="itpseq.DataSet.aafile_pattern">
<span class="sig-name descname"><span class="pre">aafile_pattern</span></span><a class="headerlink" href="#itpseq.DataSet.aafile_pattern" title="Link to this definition">¶</a></dt>
<dd><p>Pattern used to identify the amino acid files in the data_path directory.
It will use the values captured in the <cite>file_pattern</cite> regex to construct the file names.
If None, defaults to <cite>‘{lib_type}_{sample}{replicate}_aa.processed.txt’</cite></p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<p class="rubric">Examples</p>
<p>Creating a DataSet from a simple antibiotic treatment (tcx) vs no treatement (noa) with 3 replicates each (1, 2, 3).</p>
<dl>
<dt>Load a dataset from the current directory, inferring the samples automatically</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">itpseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">DataSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="n">DataSet</span><span class="p">(</span><span class="n">data_path</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span>
<span class="go">DataSet(data_path=PosixPath(&#39;.&#39;),</span>
<span class="go">   reference=Sample(noa:[1, 2, 3]),</span>
<span class="go">   samples=[Sample(noa:[1, 2, 3]),</span>
<span class="go">            Sample(tcx:[1, 2, 3], ref: noa)],</span>
<span class="go">   )</span>
</pre></div>
</div>
</dd>
<dt>Compute a standard report and export it as PDF</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">report</span><span class="p">(</span><span class="s1">&#39;my_experiment.pdf&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd>
<dt>Display a graph of the inverse-toeprints lengths for each sample</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">itp_len_plot</span><span class="p">(</span><span class="n">row</span><span class="o">=</span><span class="s1">&#39;sample&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Attributes<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>samples_with_ref</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="../api.html#itpseq.DataSet.DE" title="itpseq.DataSet.DE"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DE</span></code></a>([pos])</p></td>
<td><p>Computes the log2-FoldChange for each motif described by <cite>pos</cite> for each sample in the DataSet relative to their reference</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="../api.html#itpseq.DataSet.infos" title="itpseq.DataSet.infos"><code class="xref py py-obj docutils literal notranslate"><span class="pre">infos</span></code></a>([html])</p></td>
<td><p>Displays summary information about the dataset sequences.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<tbody>
<tr class="row-odd"><td><p><strong>itoeprint</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>itp_len_plot</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>reorder_samples</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>report</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="itpseq.DataSet.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'.'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">result_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keys</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_labels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">tuple</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'noa'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_pattern</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aafile_pattern</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#itpseq.DataSet.__init__" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="../api.html#itpseq.DataSet.DE" title="itpseq.DataSet.DE"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DE</span></code></a>([pos])</p></td>
<td><p>Computes the log2-FoldChange for each motif described by <cite>pos</cite> for each sample in the DataSet relative to their reference</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#itpseq.DataSet.__init__" title="itpseq.DataSet.__init__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__init__</span></code></a>([data_path, result_path, samples, ...])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="../api.html#itpseq.DataSet.infos" title="itpseq.DataSet.infos"><code class="xref py py-obj docutils literal notranslate"><span class="pre">infos</span></code></a>([html])</p></td>
<td><p>Displays summary information about the dataset sequences.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">itoeprint</span></code>([plot, norm, norm_range, ...])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">itp_len_plot</span></code>([ax, col, row, min_codon, ...])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">reorder_samples</span></code>(order[, validate, ...])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">report</span></code>([template, output])</p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<p class="rubric">Attributes</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">samples_with_ref</span></code></p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">toeprint_df</span></code></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
</dd></dl>

</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">iTP-Seq</a></h1>









<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2025, Thibaud Renault.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.1.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="../_sources/generated/itpseq.DataSet.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>